### PR TITLE
Optimize space consumption to O(n) in edit distance

### PR DESCRIPTION
**Pull request by MaskRay**

### CODE REVIEW

**Comment by MaskRay** [comment]
The idea is, only one row is required to calculate the 2D matrix.

`ul` for the upper left cell and `u` for the upper cell.

**Comment by radare** [comment]
same as for the other PR, optimizing is no prio when travis is red and there are no tests to ensure your optimizations are correct, because we have already reviewed and checked that code a bunch of times in the past. *focus*

**Comment by MaskRay** [comment]
This is unlikely to be changed by other commits so I won't have rebase nightmare. I am fine to wait long on this issue. I worked on this b/c I've seen complaints about using too much space in the IRC channel.

**Comment by radare** [comment]
just put it in a separate function and add an option to choose and it will be merged inmediately

**Comment by MaskRay** [comment]
Refactoring it is important b/c otherwise it just refuses many requests:

```
% radiff2 -s =true =false                                                            
Too much memory required (3.5G) to run distance diff, Use -c.                                              
similarity: 0.000         
distance: 0  
```

**Comment by radare** [comment]
please rebase

**Comment by MaskRay** [comment]
Rebased

File: libr/util/diff.c [hunk 1]
```diff
@@ -358,67 +358,50 @@ R_API bool r_diff_buffers_distance_myers(RDiff *diff, const ut8 *a, ut32 la, con
...
-R_API bool r_diff_buffers_distance_original(RDiff *d, const ut8 *a, ut32 la, const ut8 *b, ut32 lb, ut32 *distance, double *similarity) {
-	int i, j, tmin, **m;
-	ut64 totalsz = 0;
-
-	if (!a || !b || la < 1 || lb < 1)
+R_API bool r_diff_buffers_distance_original(RDiff *diff, const ut8 *a, ut32 la, const ut8 *b, ut32 lb, ut32 *distance, double *similarity) {
+	if (!a || !b)
 		return false;
 
-	if (la == lb && !memcmp (a, b, la)) {
```
Comment (radare) [inline libr/util/diff.c]: why remove that optimization?

File: libr/util/diff.c [hunk 2]
```diff
@@ -358,67 +358,50 @@ R_API bool r_diff_buffers_distance_myers(RDiff *diff, const ut8 *a, ut32 la, con
...
+	const ut8 *ea = a + la, *eb = b + lb;
+	ut32 *d, i, j;
+	// Strip prefix
+	for (; a < ea && b < eb && *a == *b; a++, b++) {}
+	// Strip suffix
+	for (; a < ea && b < eb && ea[-1] == eb[-1]; ea--, eb--) {}
+	la = ea - a;
+	lb = eb - b;
+
+	if (lb + 1 > SIZE_MAX / sizeof (ut32) || !(d = malloc ((lb + 1) * sizeof (ut32)))) {
```
Comment (radare) [inline libr/util/diff.c]: use calloc and avoid the *

Comment (radare) [reply libr/util/diff.c]: well, malloc should be fine as long as you are initializing it right after that line, and lb cant be negative and it's ok to be 0, so let's keep it like this

Comment (MaskRay) [reply libr/util/diff.c]: Switched to `	if (sizeof (ut32) > SIZE_MAX / (lb + 1) || !(d = malloc ((lb + 1) * sizeof (ut32)))) {`

Actually I am also against using `ut32` as length types.

File: libr/util/diff.c (hunk 1 above)
Comment (MaskRay) [reply libr/util/diff.c]: It's fully included in the stripping prefix/suffix optimization.

Comment (radare) [reply libr/util/diff.c]: wat, i dont see you comparing the pointers

Comment (MaskRay) [reply libr/util/diff.c]:
```c
	// Strip prefix
	for (; a < ea && b < eb && *a == *b; a++, b++) {}
	// Strip suffix
	for (; a < ea && b < eb && ea[-1] == eb[-1]; ea--, eb--) {}
```

If `a` and `b` are the same, the following stuff will take very little time so I don't bother add the `if (!la && !lb)` fast path.

**Pull request merged**